### PR TITLE
Basic rootless podman install and configuration

### PR DESCRIPTION
**Pull request by shokakucarrier**
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough

Initial commit, osbs remote host setup with rootless podman

Requires podman_user and podman_user_id

Signed-off-by: Lianghui Yu <liyu@redhat.com>

- CLOUDBLD-8567

### CODE REVIEW

File: tasks/main.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- include: certificate-setup.yml
```
Comment (qixiang) [inline tasks/main.yaml]: Doesn't this certificate-setup task install the root CA or something else? IMO, it doesn't need to be include by this role, we can use other shared role to do that.

File: tasks/podman-setup.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  ansible.builtin.shell: loginctl enable-linger
+
+- name: Set podman variable
+  set_fact:
+    podman_conf: podman.conf
+
+- name: "Install Podman on RHEL8 server"
+  package:
+    name:
+      - "@container-tools" # install module 'container-tools'
```
Comment (qixiang) [inline tasks/podman-setup.yaml]: Do we really need the whole group/module?

File: tasks/podman-setup.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      - podman-docker
```
Comment (qixiang) [inline tasks/podman-setup.yaml]: We're going to use `podman`, `podman-docker` provides the ability to run `docker` commands but talk to the podman socket, which is not used by us.

File: files/podman.conf [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# https://github.com/containers/libpod/blob/master/troubleshooting.md#5-rootless-containers-cannot-ping-hosts
+net.ipv4.ping_group_range=0 2000000
```
Comment (qixiang) [inline files/podman.conf]: I don't think we need to ping hosts from containers in our use cases, so this is unnecessary.

File: tasks/podman-setup.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    path: "/etc/{{ item }}"
+    line: "{{ user | default('cloud-user') }}:10000:65536" # to solve podman error: "setup user: invalid argument" in rootless mode
+  loop:
+    - subuid
+    - subgid
+
+- name: "Copy podman configuration"
+  copy:
+    dest: "/etc/sysctl.d/{{ podman_conf }}"
+    src: "{{ podman_conf }}"
```
Comment (qixiang) [inline tasks/podman-setup.yaml]: As above, we don't have such a requirement.

File: tasks/podman-setup.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+- name: "Start podman service"
+  ansible.builtin.systemd:
+    state: started
+    name: podman.service
```
Comment (qixiang) [inline tasks/podman-setup.yaml]: 1. Podman backend is enabled and started by `podman.socket`.
2. This starts the rootful podman while the PR says we're enabling rootless podman, we're currently not sure about whether we need rootful podman or rootless podman, but even with rootful podman, this is not sufficient, we'll need to grant the permission to unprivileged users.
3. To start rootless podman with a different user (which runs the playbooks), you have to get the login session via `machinectl`.
4. Just enable and start the service is not sufficient, you have to enable the persistent user session as well, about enabling podman backend, you can check more details in CLOUDBLD-8567.

File: tasks/podman-setup.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+- name: "Enable podman service"
```
Comment (qixiang) [inline tasks/podman-setup.yaml]: You can enable and start the service in a single task with systemd. e.g:
```
- name: Enable and start rootless podman.socket
  systemd:
    name: podman.socket
    enabled: yes 
    state: started
    scope: user
```

File: tasks/python-pip-setup.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+---
```
Comment (qixiang) [inline tasks/python-pip-setup.yaml]: I don't think we need to run pip in our cases.

File: tasks/python-setup.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (qixiang) [inline tasks/python-setup.yaml]: I think we don't need to include this in this role.

Comment (shokakucarrier) [reply tasks/python-setup.yaml]: I'm adding this is maybe in some case, we need to run python to do the locking and unlocking mechanism

File: tasks/podman-setup.yaml (hunk 6 above)
Comment (shokakucarrier) [reply tasks/podman-setup.yaml]: This script is still not ready to start rootless podman, it's pending on the decision of desgin part

File: tasks/python-setup.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+- name: Install python38 module packages
+  package:
+    name: "@python36"
```
Comment (ben-alkov) [inline tasks/python-setup.yaml]: Should be `@python38`

**Comment by ben-alkov** [comment]
@shokakucarrier; Please provide a description for this PR, including a DCO (e.g. mine is 'Signed-off-by: Ben Alkov ben.alkov@redhat.com', and the ticket number you're working from (should be CLOUDBLD-xxxx)

File: meta/main.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+  author: Lianghui Yu
+  description: Setup remote host for OSBS podman build
+  company: Red Hat, Inc.
+  license: BSD
+  min_ansible_version: 2.1
+  platforms:
+    - name: EL
+      versions:
+        - 8
+dependencies: []
```
Comment (MartinBasti) [inline meta/main.yaml]: can we pls have newline at the end

File: tasks/python-setup.yaml (hunk 9 above)
Comment (MartinBasti) [reply tasks/python-setup.yaml]: so far we don't need python

**Comment by MartinBasti** [comment]
Commit message says `rootful` podman. IMO we are using rootless according the code

File: files/podman.conf [hunk 12]
```diff
@@ -0,0 +1 @@
...
```
Comment (qixiang) [inline files/podman.conf]: I don't see a requirement for enabling the unprivileged ping at this moment, so I'd suggest not including this until we know we actually need it. And this configuration is not specific to podman, a more meaningful name like `allow_unprivileged_ping.conf` could be better.

File: tasks/main.yaml [hunk 13]
```diff
@@ -0,0 +1,32 @@
+---
+- name: Create user accounts
+  user:
+    name: "{{ podman_user }}"
+    uid: "{{ podman_user_uid }}"
+
+- name: Configure loginctl to prevent logout causing podman to exit
```
Comment (qixiang) [inline tasks/main.yaml]:
```suggestion
- name: Enable persistent user session to allow running long-running services without logged in
```

File: tasks/main.yaml [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  command: loginctl enable-linger {{ podman_user }}
+
+- name: Set podman variable
+  set_fact:
+    podman_conf: podman.conf
```
Comment (qixiang) [inline tasks/main.yaml]: As above, we can check with osbs team whether there is any use cases that using unprivileged ping during building process.

File: tasks/main.yaml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+- name: Install Podman on RHEL8 server
+  dnf:
+    name:
+      - podman
+    state: latest
+
+- name: Copy podman configuration
+  copy:
+    dest: "/etc/sysctl.d/{{ podman_conf }}"
+    src: "{{ podman_conf }}"
```
Comment (qixiang) [inline tasks/main.yaml]: As above.

File: tasks/main.yaml [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+- name: Install machinectl to start podman service
+  dnf:
+    name:
+      - systemd-container
+    state: latest
+
+- name: Enable and start rootless podman.socket
+  command: machinectl shell {{ podman_user }}@.host /bin/sh -c "systemctl --user start podman.socket"
```
Comment (qixiang) [inline tasks/main.yaml]: This only starts the podman backend but not enable it, and it's better to add comment to explain why we need to do it in this way.
```suggestion
# rootless podman backend need to be started with a login session, so
# ansible become methods ("sudo", "su" and etc) doesn't work in this
# case, become method "machinectl" is not flexible enough to fulfill
# the requirement either, we need to use command module as a workaround
- name: Enable and start rootless podman backend
  command: machinectl shell {{ podman_user }}@.host /usr/bin/systemctl --user enable --now podman.socket
```

**Review by qixiang — CHANGES_REQUESTED** [review]
Could you update the commit message to reflect the latest status? it's not rootful, no python installed, and please rebase it to trigger the lint in CI, there are some ansible-lint failures and warnings need to be fixed.

**Review by MartinBasti — APPROVED** [review]

**Review by qixiang — APPROVED** [review]
:+1:

**Comment by ben-alkov** [comment]
@shokakucarrier; Please add newlines to the end of 'meta/main.yaml' and 'tasks/main.yaml'

**Review by ben-alkov — APPROVED** [review]

**Pull request merged**